activemodel/lib/active_model/secure_password.rb

```diff
@@ -90,6 +90,7 @@ def has_secure_password(attribute = :password, validations: true)
 
         include InstanceMethodsOnActivation.new(attribute)
 
+        validations = validations.call if validations.is_a?(Proc)
         if validations
           include ActiveModel::Validations
 
```

Comment:
> e.g. Inviting a user without having to set a password for them yet so they can add it themselves later when they receive an email invitation to finish setting up their account.

Your models may be different than I'm imagining, but I think it would be good to give the Proc access to the model instance, no?  That way it could be based on e.g. `user.guest?` instead of requiring a `GuestUser` model class.

So what about reusing the `:if` / `:unless` logic that [`validates`](https://api.rubyonrails.org/v7.0/classes/ActiveModel/Validations/ClassMethods.html#method-i-validates) provides:

```diff
index 7cc207204e..bd584483bd 100644
--- a/activemodel/lib/active_model/secure_password.rb
+++ b/activemodel/lib/active_model/secure_password.rb
@@ -93,15 +93,17 @@ def has_secure_password(attribute = :password, validations: true)
         if validations
           include ActiveModel::Validations

+          validation_options = validations.is_a?(Hash) ? validations : {}
+
           # This ensures the model has a password by checking whether the password_digest
           # is present, so that this works with both new and existing records. However,
           # when there is an error, the message is added to the password attribute instead
           # so that the error message will make sense to the end-user.
-          validate do |record|
+          validate(validation_options) do |record|
             record.errors.add(attribute, :blank) unless record.public_send("#{attribute}_digest").present?
           end

-          validate do |record|
+          validate(validation_options) do |record|
             if challenge = record.public_send(:"#{attribute}_challenge")
               digest_was = record.public_send(:"#{attribute}_digest_was") if record.respond_to?(:"#{attribute}_digest_was")

@@ -111,8 +113,8 @@ def has_secure_password(attribute = :password, validations: true)
             end
           end

-          validates_length_of attribute, maximum: ActiveModel::SecurePassword::MAX_PASSWORD_LENGTH_ALLOWED
-          validates_confirmation_of attribute, allow_blank: true
+          validates_length_of attribute, **validation_options, maximum: ActiveModel::SecurePassword::MAX_PASSWORD_LENGTH_ALLOWED
+          validates_confirmation_of attribute, **validation_options, allow_blank: true
         end
       end
     end
```

Reply:
Thanks for your reply. 

So the give your more context, My app allows _someone_ to register. You have to fill in the whole registration form (password included) at that stage.
Once you are all setup, The user is created along with a "company" they will belong to.

After that, they can now decide to invite users to their company. 

When you do that, it's still the same User model, I don't have the notion of `GuestUser`, all I am doing, or wanting to do at that point, is bypassing validations for password when my user is being created "internally". 
What I was wanting to do was to set a flag which would turn a variable to false so the validations is not triggered 

`has_secure_password validations: Proc.new { |user| user.skip_password_validations? }`

That's for the context. 

Now looking at your comment, this makes a lot of sense, and definitely looks like a better implementation than what I have done. I am happy to update my test and the code to reflect that 👍 

On the last 2 lines you changed, does it matter to have the `**validation_options` at the end of the line rather than before the other rules ? _just a preference from me to have it at the end_

Reply:
This has been updated.
I also took the opportunity to rename my test model as it wasn't reflecting what I was trying to do.

Let me know what you think.